RavenDB-17793 prefixed sharding

ppekrol opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17793

ppekrol commented:
@karmeli87 @ayende Would like to have a chat about this in context of:
- how to define it
- how to validate the values
- how to handle resharding

ayende commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -3429,7 +3429,7 @@ public ShardingConfiguration ReadShardingConfiguration<TTransaction>(Transaction
             using (var raw = ReadRawDatabaseRecord(context, name))
             {
-                return raw.Sharding.Value;
+                return raw.Sharding.MaterializedConfiguration;
```
We need to ensure that this is something that can only be configured once (the prefixes setup, the actual distribution inside can change).

aviv86 committed:
RavenDB-17793
- Can now specify different sharding topologies based on document prefixes.
- Extract the logic for translating a document id to a shard number to a single location
- Refactoring how we hold sharding data inside the database record

aviv86 committed:
RavenDB-17793 : use different ranges for prefixed sharding :
if item id is covered in prefixed-sharding settings, then it's bucket will be computed as follows : GetBucket(id) + (1_000_000 * position-in-prefixed-settings)

aviv86 committed:
RavenDB-17793 : fix failing backup/smuggler tests : SingleShardDatabaseSmuggler.ExecuteAsync should not return task without awaiting it inside a using statement

aviv86 committed:
RavenDB-17793 : ShardedCompareExchangeActions : use docId instead of atomic-guard key to determine the shard number

aviv86 commented:
rebased, fixed errors and made a few changes:

- each prefix defined in `record.Sharding.Prefixed` will get a different range of 1024 * 1024 buckets, based on it's position in `record.Sharding.Prefixed`

- e.g. if we have :
```
Prefixed = 
{ 
	'eu/' 	: [shard 0], 
	'asia/' : [shard 1], 
	'us/' 	: [shard 0, shard 2] 	
} 
```
then the bucket ranges will look like :

```
// default bucket ranges
shard 0 : [0, 333K] 
shard 1 : [333K, 666K]
shard 2 : [666K, 1M] 

// prefixed ranges
shard 0 : [1M, 2M] // 'eu/' range
shard 1 : [2M, 3M]  // 'asia/' range
shard 0 : [3M, 3.5M]  // 'us/' range
shard 2 : [3.5M, 4M]  // 'us/' range
```

- instead of storing the 'prefixed' ranges separately in `record.Sharding.Prefixed`, now they are placed in `record.Sharding.BucketRanges` after all 'regular' ranges 


- to determine the bucket of an id we use `GetBucket(id)` and if the id is prefixed we add 1024 * 1024 * `position-in-prefixed-settings` to it

ayende commented:
How do I move a bucket in the `us/` prefix from one shard to another?

ayende commented on src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs:
```diff
@@ -159,14 +160,32 @@ internal static ByteStringContext.Scope ExtractIdFromKeyAndGenerateBucketAndEtag
-        var bucket = ShardHelper.GetBucket(idPtr, idSize);
+        var bucket = ShardHelper.GetBucketFor(idPtr, idSize);
+
+        var database = tx.Owner as ShardedDocumentDatabase;
```
Can this *not* be the case?

aviv86 replied:
it can be a regular DocumentDatabase

ayende commented on src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs:
```diff
@@ -159,14 +160,32 @@ internal static ByteStringContext.Scope ExtractIdFromKeyAndGenerateBucketAndEtag
+        var prefixedConfiguration = database?.ReadShardingState().Prefixed;
+        if (prefixedConfiguration != null)
+        {
+            var idAsStr = Encoding.UTF8.GetString(idPtr, idSize);
```
didn't we start with a string in the caller?

aviv86 replied:
no, we start with a TableValueReader and id position in value reader

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -872,53 +872,82 @@ private void BeforeAppendToRaftLog(ClusterOperationContext ctx, CommandBase cmd)
+                for (int i = 0; i < prefixedSettings.Count; i++)
+                {
+                    var shards = prefixedSettings[i].Shards;
+                    var start = ShardHelper.NumberOfBuckets << i;
```
This multiples, not add.
So you have `0.. 1M, 1M .. 2M, 4M..5M, 8M..9M`, no?

aviv86 replied:
fixed

ayende commented on src/Raven.Server/Utils/ShardHelper.cs:
```diff
@@ -194,11 +162,82 @@ public static IEnumerable<string> GetShardNames(string databaseName, int shardsC
             {
-                if (bucket < ranges[i + 1].BucketRangeStart)
+                int bucketRangeStart = ranges[i].BucketRangeStart;
+                if ((bucketRangeStart >> 20) != prefixRange)
```
This approach limits us to ~12 ranges, no?
But if we use additions it will have ~4K

karmeli87 replied:
yes, we should be able to support up to 4k prefixes

aviv86 replied:
fixed

karmeli87 commented:
Move bucket 3,000,001 (`us/` prefix range is approximately 3,000,000 - 4,000,000) from shard 0 to shard 2

aviv86 committed:
RavenDB-17793 : change Sharding.Prefixed such that it will only contain shard numbers by prefix (instead of List<ShardBucketRange> by prefix) and place the prefixed ranges at the end of Sharding.BucketRanges

aviv86 committed:
RavenDB-17793 : sharded backup and restore should take into account Sharding.Prefix configuration

aviv86 committed:
RavenDB-17793 : address some PR comments and change Prefixed configuration to List<PrefixedShardingSetting>, were PrefixedShardingSetting has a bucketRangeStart, prefix and shard numbers

aviv86 committed:
RavenDB-17793 : fixes after rebase

aviv86 committed:
RavenDB-17793 : Sharding.Prefixed : handle updates (assert no docs starting with prefix on add/delete), move all relevant code from ServerStore to ShardedDatabaseContext, add tests

aviv86 committed:
RavenDB-17793 : fixes after rebase

aviv86 committed:
RavenDB-17793 : fixes after rebase

aviv86 committed:
RavenDB-17793 : fix bug in admin/databases PUT : use sharded db names when checking if we should unload and lock database for recreate indexes

aviv86 committed:
RavenDB-17793 : handle prefixed sharding update outside of FillPrefixedSharding, because we don't want to do async calls (AssertNoDocsStartingWith) inside a write tx (FillPrefixedSharding is called from ServerStore.BeforeAppendToRaftLog)

ppekrol commented on src/Raven.Client/Documents/Commands/GetDocumentsCommand.cs:
```diff
@@ -35,6 +35,8 @@ public class GetDocumentsCommand : RavenCommand<GetDocumentsResult>
         private readonly string _exclude;
         private readonly string _startAfter;
 
+        internal string _database;
```
this does not look correct?

aviv86 replied:
fixed

ppekrol commented on src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.cs:
```diff
@@ -139,6 +150,242 @@ private void UpdateConfiguration(Dictionary<string, string> settings)
             Configuration = DatabasesLandlord.CreateDatabaseConfiguration(ServerStore, DatabaseName, settings);
         }
 
+        public static void FillShardingConfiguration(ServerStore serverStore, AddDatabaseCommand addDatabase, ClusterTopology clusterTopology)
```
does not look like an appropriate place for this?

aviv86 replied:
moved to `ServerStore.Sharding`

ppekrol commented on src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.cs:
```diff
@@ -139,6 +150,242 @@ private void UpdateConfiguration(Dictionary<string, string> settings)
+            }
+        }
+
+        private static void AddPrefixedBucketRange(PrefixedShardingSetting setting, int rangeStart, ShardingConfiguration shardingConfiguration)
```
why this is here?

aviv86 replied:
moved to `ServerStore.Sharding`

aviv86 committed:
RavenDB-17793 : address PR comments and fix errors after rebase

aviv86 committed:
RavenDB-17793 applied prefixed sharding

aviv86 committed:
RavenDB-17793 : fixes after rebase

aviv86 committed:
RavenDB-17793 : 'BucketAndEtag' storage indexes should take into account prefixed-sharding settings when computing the bucket

aviv86 committed:
RavenDB-17793 : handle moving bucket of prefixed bucket range + basic test

aviv86 committed:
RavenDB-17793 : AttachmentsStorage.GetBucketFromAttachmentKey should take into account Prefix settings

aviv86 committed:
RavenDB-17793 : block prefix sharding usage, use it only in tests

aviv86 committed:
RavenDB-17793 : fixes after rebase

karmeli87 commented:
@ppekrol @ayende @aviv86 
I'm merging this since there are a lot of changes that we want to have in our code base regardless to the feature itself.
Addressing any comment will be done on a different PR
